### DISCUSSION TITLE
VR Validation in v5

### DISCUSSION THREAD

**Discussion by MichaelM223** [fo-dicom/fo-dicom#1384] [Q&A] [answered]
Hi all,

I've got a question about the new way fo-dicom v5 handles VR validation; v4 used to disable validation on cloning of a dicom file but v5 does not. The only option I see to disable VR validation is globally via a DicomSetupBuilder instance, something I do not wish to do.

Is there a way to locally toggle validation of a dataset instead of using a DicomSetupBuilder instance?

**Comment by MichaelM223** [comment]
I've fixed it for now with the following, but it feels a bit hacky/misuse of the DicomSetupBuilder:

```csharp
private DicomSetupBuilder _dicomSetupBuilder;

public BaseDicomAnonymizer(SecurityProfile profile = null) : base(profile) 
{
    _dicomSetupBuilder = new DicomSetupBuilder();
}

public new DicomDataset Anonymize(DicomDataset dataset)
{
    _dicomSetupBuilder.SkipValidation();
    var clone = dataset.Clone();
    _dicomSetupBuilder.DoValidation();
    AnonymizeInPlace(clone);
    return clone;
}

public new DicomFile Anonymize(DicomFile file)
{
    _dicomSetupBuilder.SkipValidation();
    var clone = file.Clone();
    _dicomSetupBuilder.DoValidation();
    AnonymizeInPlace(clone);
    return clone;
}
```

Wouldn't it be better to make line 13 in `DicomValidation.cs` `public` instead of `internal` so it can be changed without using a setupbuilder instance?
```csharp
internal static bool PerformValidation { get; set; } = true;
```

**Comment by gofal** [comment]
what makes you assume that in version 5 the clone method does validation?

https://github.com/fo-dicom/fo-dicom/blob/246f14e8adc3f7b9c8aa7c1fc2075cd71ca24b35/FO-DICOM.Core/DicomDatasetExtensions.cs#L22-L26

Clone is just copying the dataset and passing the parameter "validate: false".

**Reply by MichaelM223** [reply]
Sorry for the late reply; I should have specified the issue was for cloning of a DicomFile instance. (in my example the workaround for the DicomDataset cloning is unnecessary as you pointed out)

The `.Add()` methods used in `DicomFile.Clone()` will trigger validation and can only be worked around on a per file basis by (mis)using `DicomSetupBuilder`.

https://github.com/fo-dicom/fo-dicom/blob/246f14e8adc3f7b9c8aa7c1fc2075cd71ca24b35/FO-DICOM.Core/DicomFileExtensions.cs#L17-L24

fo-dicom v4 did disable validation on cloning of a `DicomFile` instance:

https://github.com/fo-dicom/fo-dicom/blob/4.0/DICOM/DicomFileExtensions.cs#L16-L27

**Comment by MichaelM223** [comment, marked as answer]
Created pull request for this issue #1466
